builtin/game/hud.lua

```diff
@@ -0,0 +1,238 @@
+--[[
+Register function to easily register new builtin hud elements
+Contains:
+  elem_def  the HUD element definion which can be changed with hud_replace_builtin
+  event     (optional) for an additional eventname on which update_element is called
+            ("hud_changed" and "properties_changed" will always be used.)
+  hud_change(id, player)
+            (optional) a function to change the element after it has been updated
+  show_elem(player, flags, id)
+            (optional) a function to decide if the element should be shown to a player
+]]--
+local registered_elements = {}
+local update_events = {}
+local function register_builtin_hud_element(name, def)
+	registered_elements[name] = def
+	if def.event then
+		local events = update_events[def.event]
+		if events then
+			table.insert(events, name)
+		else
+			update_events[def.event] = {name}
+		end
+	end
+end
+
+-- Stores HUD ids for all players
+local hud_ids = {}
+
+-- Updates one element
+-- In case the element is already added, it does only call the hud_change function from
+-- registered_elements. (To completely update the element call player:hud_remove(id) first.)
+local function update_element(player, player_hud_ids, elem_name, flags)
+	local def = registered_elements[elem_name]
+	local id = player_hud_ids[elem_name]
+
+	if def.show_elem and not def.show_elem(player, flags, id) then
+		if id then
+			player:hud_remove(id)
+		end
+		return
+	end
+
+	if not id then
+		id = player:hud_add(def.elem_def)
+		player_hud_ids[elem_name] = id
+	end
+
+	if def.hud_change then
+		def.hud_change(player, id)
+	end
+end
+
+-- Updates all elements
+local function update_hud(player)
+	local flags = player:hud_get_flags()
+	local playername = player:get_player_name()
+	local player_hud_ids = hud_ids[playername]
+	if not player_hud_ids then
+		player_hud_ids = {}
+		hud_ids[playername] = player_hud_ids
+	end
+	for elem_name, _ in pairs(registered_elements) do
+		update_element(player, player_hud_ids, elem_name, flags)
+	end
+end
+
+local function player_event_handler(player, eventname)
+	assert(player:is_player())
+
+	if eventname == "hud_changed" or eventname == "properties_changed" then
+		update_hud(player)
+		return true
+	end
+
+	-- Custom events
+	local to_update = update_events[eventname]
+	if to_update then
+		local flags = player:hud_get_flags()
+		local playername = player:get_player_name()
+		local player_hud_ids = hud_ids[playername]
+		if not player_hud_ids then
+			player_hud_ids = {}
+			hud_ids[playername] = player_hud_ids
+		end
+		for _, elem_name in ipairs(to_update) do
+			update_element(player, player_hud_ids, elem_name, flags)
+		end
+		return true
+	end
+
+	return false
+end
+
+-- Returns true if successful, otherwise false,
+-- but currently the return value is not part of the lua API.
+function core.hud_replace_builtin(elem_name, definition)
+	if type(definition) ~= "table" or
+			(definition.type or definition.hud_elem_type) ~= "statbar" then
+		return false
```

Comment:
should be an `assert` imo

Reply:
Oh, I also forgot to remove the statbar check when I rebased and the minimap element got added.

Reply:
I replaced it with an assert, but this is now a change in the behavior of `minetest.hud_replace_builtin` since the unrefactored code just returned `false` in this case.